### PR TITLE
Growable Buffer header

### PR DESCRIPTION
**Pull request by ManasviGoyal**
New header-only implementation of Growable Buffer using panels. Contains the following :-

- [x]  GrowableBuffer.h
- [x]  Unit Tests for GrowableBuffer
- [x]  Documentation of GrowableBuffer.h

### CODE REVIEW

**Review by jpivarski — APPROVED** [review]
Great! The location of this file will move and it the class will get new methods, but this is a good checkpoint.

To get the tests to pass, merge with `main`. @ioanaif's #1533 fixes the pre-commit issues that you see. Once it does pass tests, you can squash-and-merge it yourself.

**Comment by ManasviGoyal** [comment]
> Great! The location of this file will move and it the class will get new methods, but this is a good checkpoint.
> 
> To get the tests to pass, merge with `main`. @ioanaif's #1533 fixes the pre-commit issues that you see. Once it does pass tests, you can squash-and-merge it yourself.

Sure. Thanks. I have added the function that is needed for use in Array Builder. So, it should be able to work with Array Builders and the new Layout Builder. New methods can be added as and when they are needed.

**Comment by jpivarski** [comment]
At some point, we should be able to remove the old GrowableBuffer, too.

Is that currently dead code? (I.e. if you just remove the old GrowableBuffer, can it compile?) When it is possible to remove, let's do that right away so that we don't end up with a name conflict somehow.

Fortunately, GrowableBuffer does not have a pybind11 interface, so there won't be anything in the src/python directory to dismantle.

File: src/awkward/_v2/cpp-headers/GrowableBuffer.h [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    /// Used to change the data type of buffer content from `FROM_PRIMITIVE`
+    /// to `TO_PRIMITIVE` for building arrays.
+    template<typename FROM_PRIMITIVE>
+    static GrowableBuffer<PRIMITIVE>
+    fill_from(GrowableBuffer<FROM_PRIMITIVE>& from_buffer) {
+      auto length = from_buffer.length();
+      auto ptr = std::unique_ptr<PRIMITIVE>(new PRIMITIVE[length]);
+      PRIMITIVE* rawptr = ptr.get();
+      int64_t k = 0;
+      for (size_t i = 0;  i < from_buffer.ptr_.size(); i++) {
```
Comment (ianna) [inline src/awkward/_v2/cpp-headers/GrowableBuffer.h]: I think, you need to add accessors to private members: ptr_ and length_

**Review by ianna — CHANGES_REQUESTED** [review]
Please, see my comment about accessing private members from a static function. Thanks.

File: src/awkward/_v2/cpp-headers/GrowableBuffer.h (hunk 1 above)
Comment (ManasviGoyal) [reply src/awkward/_v2/cpp-headers/GrowableBuffer.h]: Do we need to expose the length_ (and ptr_) of every panel? As currently, we already have a method that returns accumulated length.

Comment (jpivarski) [reply src/awkward/_v2/cpp-headers/GrowableBuffer.h]: Isn't the fact that GrowableBuffer is implemented by panels an internal detail that you don't want other code depending on? (For instance, if we find in the future that some mix of panels and reallocation is ideal... not that I think that will happen.) Exposing either `ptr_` or `length_` would reveal the panels.

Comment (ManasviGoyal) [reply src/awkward/_v2/cpp-headers/GrowableBuffer.h]: Yes, I agree. Exposing `ptr_` and `length_` should be avoided.

File: src/awkward/_v2/cpp-headers/GrowableBuffer.h [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        for (int64_t j = 0; j < from_buffer.length_[i]; j++) {
+          rawptr[k] = static_cast<PRIMITIVE>(from_buffer.ptr_[i].get()[j]);
+          k++;
+        }
+      }
+      return GrowableBuffer<PRIMITIVE>(length, std::move(ptr), length, length);
+    }
```
Comment (ianna) [inline src/awkward/_v2/cpp-headers/GrowableBuffer.h:260]:
```suggestion
    /// @brief Takes one (possibly multi-panels) GrowableBuffer<FROM_PRIMITIVE>
    /// and makes another (one panel) GrowableBuffer<TO_PRIMITIVE>.
    ///
    /// Used to change the data type of buffer content from `FROM_PRIMITIVE`
    /// to `TO_PRIMITIVE` for building arrays.
    template<typename FROM_PRIMITIVE>
    static GrowableBuffer<PRIMITIVE>
    fill_from(GrowableBuffer<FROM_PRIMITIVE>& from_buffer) {
      auto length = from_buffer.length();
      auto ptr = std::unique_ptr<PRIMITIVE>(new PRIMITIVE[length]);
      PRIMITIVE* rawptr = ptr.get();
      int64_t k = 0;
      for (size_t i = 0;  i < from_buffer.ptr_.size(); i++) {
        for (int64_t j = 0; j < from_buffer.length_[i]; j++) {
          rawptr[k] = static_cast<PRIMITIVE>(from_buffer.ptr_[i].get()[j]);
          k++;
        }
      }
      return GrowableBuffer<PRIMITIVE>(length, std::move(ptr), length, length);
    }
```

File: src/awkward/_v2/cpp-headers/GrowableBuffer.h [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
```
Comment (ianna) [inline src/awkward/_v2/cpp-headers/GrowableBuffer.h]:
```suggestion
    /// @brief Takes this (possibly multi-panels) GrowableBuffer<PRIMITIVE>
    /// and makes another (one panel) GrowableBuffer<TO_PRIMITIVE>.
    ///
    /// Used to change the data type of buffer content from `PRIMITIVE`
    /// to `TO_PRIMITIVE` for building arrays.
    template<typename TO_PRIMITIVE>
    GrowableBuffer<TO_PRIMITIVE>
    copy_as() {
      auto len = length();
      auto ptr = std::unique_ptr<TO_PRIMITIVE>(new TO_PRIMITIVE[len]);
      TO_PRIMITIVE* rawptr = ptr.get();
      int64_t k = 0;
      for (size_t i = 0;  i < ptr_.size(); i++) {
        for (int64_t j = 0; j < length_[i]; j++) {
          rawptr[k] = static_cast<TO_PRIMITIVE>(ptr_[i].get()[j]);
          k++;
        }
      }
      return GrowableBuffer<TO_PRIMITIVE>(len, std::move(ptr), len, len);
    }
```

**Review by ianna — CHANGES_REQUESTED** [review]
Ok, my compiler does not like that the static function accesses the private members. Here is an alternative implementation.

**Review by ianna — APPROVED** [review]
This looks good! Thanks. Please, merge it when the tests pass.

**Comment by jpivarski** [comment]
@all-contributors please add @ManasviGoyal for code

**Pull request merged**